ethereum/core/src/main/java/org/hyperledger/besu/ethereum/MainnetBlockValidator.java

```diff
@@ -157,53 +160,50 @@ public BlockProcessingResult validateAndProcessBlock(
           .ifPresentOrElse(
               synchronizer -> synchronizer.healWorldState(ex.getMaybeAddress(), ex.getLocation()),
               () ->
-                  handleAndLogImportFailure(
+                  handleFailedBlockProcessing(
                       block,
                       new BlockProcessingResult(Optional.empty(), ex),
-                      shouldRecordBadBlock));
+                      // Do not record bad black due to missing data
+                      false));
       return new BlockProcessingResult(Optional.empty(), ex);
     } catch (StorageException ex) {
       var retval = new BlockProcessingResult(Optional.empty(), ex);
-      handleAndLogImportFailure(block, retval, shouldRecordBadBlock);
+      // Do not record bad block due to a local storage issue
+      handleFailedBlockProcessing(block, retval, false);
       return retval;
     } catch (Exception ex) {
+      // Wrap checked autocloseable exception from try-with-resources
       throw new RuntimeException(ex);
     }
   }
 
-  private void handleAndLogImportFailure(
-      final Block invalidBlock,
+  private void handleFailedBlockProcessing(
+      final Block failedBlock,
       final BlockValidationResult result,
       final boolean shouldRecordBadBlock) {
     if (result.causedBy().isPresent()) {
+      // Block processing failed exceptionally, we cannot assume the block was intrinsically invalid
       LOG.info(
-          "Invalid block {}: {}, caused by {}",
-          invalidBlock.toLogString(),
+          "Failed to process block {}: {}, caused by {}",
+          failedBlock.toLogString(),
           result.errorMessage,
           result.causedBy().get());
       LOG.debug("with stack", result.causedBy().get());
     } else {
       if (result.errorMessage.isPresent()) {
-        LOG.info("Invalid block {}: {}", invalidBlock.toLogString(), result.errorMessage);
+        LOG.info("Invalid block {}: {}", failedBlock.toLogString(), result.errorMessage);
       } else {
-        LOG.info("Invalid block {}", invalidBlock.toLogString());
+        LOG.info("Invalid block {}", failedBlock.toLogString());
+      }
+
+      if (shouldRecordBadBlock) {
+        // Result.errorMessage should not be empty on failure, but add a default to be safe
+        String description = result.errorMessage.orElse("Unknown cause");
+        final BadBlockCause cause = BadBlockCause.fromValidationFailure(description);
+        badBlockManager.addBadBlock(failedBlock, cause);
+      } else {
+        LOG.debug("Invalid block {} not added to badBlockManager ", failedBlock.toLogString());
```

Comment:
```suggestion
        LOG.debug("Failed block {} not added to badBlockManager ", failedBlock.toLogString());
```

Reply:
The block here actually is invalid (failed validation), so I think its useful to label it that way.